### DISCUSSION TITLE
How to check  for "POWER" IR code as soon as waking up from deep sleep?

### DISCUSSION THREAD

**Discussion by probonopd** [Arduino-IRremote/Arduino-IRremote#1088] [Q&A]
Hello, what is a working approach to implementing a device that can go to deep sleep, wake up when a pin state change occurs on the IR pin, and then immediately decodes the IR code that has caused the pin state change?

I want to do this on an ESP32. The device should deep sleep (it is already doing this right now), and wake up when the remote is pressed (it is also already doing this right now), but only if the correct code triggered the device to wake up, or else immediately go back to deep sleep before doing anything else, such as connecting WLAN (it is not doing this yet, and this is what I am looking for help with).

How can I ensure that the first thing the device does when a pin state change woke it up is decode the IR code? Can I put the IR decoding into an ISR, or is this not the way to go?

Any ideas appreciated. Thanks!

**Comment by ArminJo** [comment]
Be aware, that the ESP32 runs a RTOS and therefore there may be many problems with fast wakeups and ISR's. You must look in the RTOS manual how to implement your request, this is only an IR library discussion panel 😞 .

**Reply by probonopd** [reply]
Yes, I was just thinking: "Probably many people have been using this library to do just that!", as many gadgets using infrared want to go to deep sleep (and wake up from it again).

**Reply by bengtmartensson** [reply]
> many gadgets using infrared want to go to deep sleep (and wake up from it again).

IFAIAA, they in general(?) deploy a co-processor for this task.

**Reply by probonopd** [reply]
Hello @bengtmartensson, amazing how quickly you find any discussion involving infrared remote control ;-)

I think I have found [two possible approaches](https://github.com/probonopd/esp32-i2s-webradio/issues/4) that would not require additional hardware, what do you think about them?

**Reply by bengtmartensson** [reply]
> Hello @bengtmartensson, amazing how quickly you find any discussion involving infrared remote control ;-)

I have been hospitalized for a few days, released today. Otherwise it would have been quicker... ;-).

**Reply by probonopd** [reply]
Get well soon!

**Comment by probonopd** [comment]
Possibly I need to use a **[Deep Sleep Wake Stub.](https://github.com/espressif/esp-idf/tree/3592638773/examples/system/deep_sleep_wake_stub)**, which runs immediately when the device is woken up from deep sleep by a pin state change on the IR pin.

My plan is to have the Deep Sleep Wake Stub measure the timings of the IR signal and put them into an array, which I can later (after the Deep Sleep Wake Stub already has exited and the device has fully booted) interpret using Arduino-IRremote.

Now the question becomes how get a Deep Sleep Wake Stub to do that...  https://github.com/espressif/arduino-esp32/discussions/7829

**Reply by ArminJo** [reply]
What is the startup time after a deep sleep? The preamble time of NEC protocol is 9 or 12 ms. so it might be possible to just inject the first one or two timings into IRRemote. 
Can you restrict your remote protocol to NEC?

**Reply by probonopd** [reply]
Unfortunately their documentation doesn't specify the number of milliseconds it takes until the deep sleep wake stub starts running after the ESP32 wakes from deep sleep. I hope that it is quick enough so that the preamble is not over yet.

**Comment by gm-jiang** [comment]
@probonopd @ArminJo 
| Chips      | CPU Clock Source | Default div | CPU clock (MHz) | Cycles | Time (ms) |
| ---------- | ---------------- | ----------- | --------------- | ------ | --------- |
| ESP32      | XTAL 40MHz       | 1           | 40              | 28736  | 0.71      |
| ESP32-ECO3 | XTAL 40MHz       | 1           | 40              | 33492  | 0.83      |
| ESP32-S2   | XTAL 40MHz       | 2           | 20              | 32637  | 1.63      |
| ESP32-S3   | XTAL 40MHz       | 2           | 20              | 149815 | 7.49      |
| ESP32-C3   | XTAL 40MHz       | 2           | 20              | 254682 | 12.73     |

* **Cycles**: Deep-sleep wakeup to RTC Wake stub

**Reply by ArminJo** [reply]
@probonopd 
So you can wait after the wake up for the timer ISR to change the state from IR_REC_STATE_IDLE to IR_REC_STATE_MARK or call void IRReceiveTimerInterruptHandler() manually to do so. 
Then you can add the "lost ticks" with addTicksToInternalTickCounter() or addMicrosToInternalTickCounter() to compensate for startup time.

**Reply by probonopd** [reply]
Thank you @gm-jiang and @ArminJo. A typical NEC IR remote control signal has a 9ms leading pulse burst followed by a 4.5ms space. So the interesting part of the signal starts after 13.5ms. So it seems like we should have enough time for the interesting part of the IR signal to begin, and should be able to measure the timestamps of pin changes in the ESP32 deep sleep wake stub. This is what I am currently [struggling with](https://github.com/espressif/arduino-esp32/discussions/7829#discussioncomment-5182269). 
Once the deep sleep wake stub gives us the timings, we should then be able to convert them to ms, and process them further using the Arduino-IRremote library.
